kore/src/Kore/Equation/Application.hs

```diff
@@ -148,11 +156,17 @@ attemptEquation
     -> simplifier (AttemptEquationResult variable)
 attemptEquation sideCondition termLike equation =
     whileDebugAttemptEquation' $ runExceptT $ do
-        let Equation { left } = equationRenamed
-        matchResult <- match left termLike & whileMatch
+        let Equation { left, argument, antiLeft } = equationRenamed
         (equation', predicate) <-
-            applyMatchResult equationRenamed matchResult
-            & whileApplyMatchResult
+            if isTop argument
+                then do
+                    matchResult <- match left termLike & whileMatch
+                    applyMatchResult equationRenamed matchResult
+                        & whileApplyMatchResult
+                else
+                    applyAndSelectMatchResult
+                    =<< simplifyArgumentWithResult argument antiLeft
+                    =<< whileMatch (match left termLike)
```

Comment:
Can we refactor this so that it is clear that both branches call `match` first? And, let's wrap the entire block in `whileMatch`, because `simplifyArgumentWithResult` is a sort of generalized matching procedure.